### ISSUE TITLE
BUG: Cannot close preview from QuickLook

### ISSUE THREAD

**Issue by nanoraptor** [Flow-Launcher/Flow.Launcher.Plugin.QuickLook#3]
https://github.com/user-attachments/assets/02ec3f23-0f1d-4468-b4d4-8fae3a562463

When I close the preview from the QuickLook, the preview panel is not going away and is very annoying when I hover my cursor over the results. It is prettty self explanatory and reproducable.

**Comment by jjw24** [comment]
Do you have always preview turned on?

**Comment by jjw24** [comment]
You can press ESC to close the QL preview window instead of using the mouse.

**Comment by nanoraptor** [comment]
> Do you have always preview turned on?

No

**Comment by nanoraptor** [comment]
> You can press ESC to close the QL preview window instead of using the mouse.

 Yea, It worked but when I control something on the QuickLook window using mouse (e.g. Media Playback) I find it easier to click the cross button rather than pressing Esc key.

**Comment by jjw24** [comment]
Yeah I understand.

I am afk, but if I remember correctly this is where QL doesn't provide a way to check whether its window is opened or closed, so flow's own code can only manage properly the state when user uses esc/preview hotkey to toggle open/close, and under the hood flow manually tracks the state. This is where if you close the window with the cross, flow won't know that the window is closed and still thinks preview is still on and when you move the mouse afterwards it switches to the preview of another result.

**Comment by nanoraptor** [comment]
Is it possible to introduce a shortcut that enables QL only while the keybinding is being pressed [#4](https://github.com/Flow-Launcher/Flow.Launcher.Plugin.QuickLook/issues/4).

**Comment by jjw24** [comment]
Yes we can either make it default behaviour where if third party preview is used don't use the built-in, or make this an option to use one or both. No need for a keyboard shortcut.

This won't solve the root cause of your problem though because the actual desired behaviour is when always preview is off and you close the preview window with the cross, no preview should be further triggered with the mouse or up down arrow until you reactivate it again (which is the behaviour of closing with ESC/preview hotkey). I guess though in light of what's beyond our control, implementing issue 4 at least helps alleviate your UX problem.

**Comment by swishlabby** [comment]
> Yes we can either make it default behaviour where if third party preview is used don't use the built-in, or make this an option to use one or both. No need for a keyboard shortcut.
> 
> This won't solve the root cause of your problem though because the actual desired behaviour is when always preview is off and you close the preview window with the cross, no preview should be further triggered with the mouse or up down arrow until you reactivate it again (which is the behaviour of closing with ESC/preview hotkey). I guess though in light of what's beyond our control, implementing issue 4 at least helps alleviate your UX problem.

Even after pressing the ESC key, If I hover my cursor to a non explorer result and back to the explorer result, preview window pops up. It is only really turned off by Alt+P.

**Comment by nanoraptor** [comment]
> Yes we can either make it default behaviour where if third party preview is used don't use the built-in, or make this an option to use one or both. No need for a keyboard shortcut.
> 
> This won't solve the root cause of your problem though because the actual desired behaviour is when always preview is off and you close the preview window with the cross, no preview should be further triggered with the mouse or up down arrow until you reactivate it again (which is the behaviour of closing with ESC/preview hotkey). I guess though in light of what's beyond our control, implementing issue 4 at least helps alleviate your UX problem.

Even after pressing the ESC key, If I hover my cursor to a non explorer result and back to the explorer result, preview window pops up. It is only really turned off by Alt+P.

**Comment by jjw24** [comment]
If you use only the preview hotkey, assuming it's your Alt + P, it should turn on off properly. I just tested using only preview hotkey and escape, the preview turns off until you toggle with the hotkey again.

**Comment by nanoraptor** [comment]
https://github.com/user-attachments/assets/d96fae90-0a0e-4178-ae00-1e8b3ff1f815

As you can see when I close quicklook using ESC key when I hover to a non explorer result the preview turns back on. 
I pressed the ESC key at 0:17

**Comment by nanoraptor** [comment]
> If you use only the preview hotkey, assuming it's your Alt + P, it should turn on off properly. I just tested using only preview hotkey and escape, the preview turns off until you toggle with the hotkey again.

That's what I was telling too. I can only use the alt+P to close the QL from coming up again. But when I interact with the QL window, I need to click flow launcher to get back the focus on the window then press alt+P to remove preview(Pressing ESC does not gurantee to turn preview off as you can see in my previous comment). This should not happen I should be just able to click the esc and close the quicklook window. It is an unwanted process.

**Comment by jjw24** [comment]
Yeah I think you are right, in my testing looks like only the preview hotkey guarantees properly closing of preview.

Escape doesn't properly turn off due to the fact that it tells QL to close the window, but flow doesn't know that the QL window is closed because there is no feed back to flow or an ability for flow to check QL's window status. This means flow thinks preview is still on and you are just switching to previewing other results. Under the hood the switching preview code is still called every time you hover to a result, it's not showing the QL window because the call sent to QL is for switching, which is ignored by QL process because it's already close the preview window. When you are hovering to a result that can only be previewed by the built-in preview, the switch is successful because it switches to the built-in, that's why you still get the built-in preview.

In a nutshell, if QL can provide a command to return the status of the preview window is opened/closed, this whole experience will be significantly better.

**Comment by nanoraptor** [comment]
So, the feature cannot be implemented properly???? QL is not updated in a long time. It would've been better to integrate powertoys peek in the plugin as it is constantly getting updated. However I prefer QL more. Is it possible for you/other devs to update this specific part of the QL's and upload it especially for Flow launcher. If this issue is solved then there is nothing much to change in QL.

**Comment by jjw24** [comment]
I think QL is a great app and I'd much rather to encourage these changes (save position + window open status feedback) over at repo, or at least get a reply from the authors that they won't add/willing to accept PRs.

**Comment by nanoraptor** [comment]
> I think QL is a great app and I'd much rather to encourage these changes (save position + window open status feedback) over at repo, or at least get a reply from the authors that they won't add/willing to accept PRs.

I do think it's a great app. But, I don't think the authors are gonna help in this one. They haven't been replying to issues for about a year. If you look at the repo it is full of feature requests and enhancements and the plugin support.

**Comment by nanoraptor** [comment]
Don't know why but my default preview key went from alt + p to f1 key. Is there a way I can change that?

**Comment by nanoraptor** [comment]
@jjw24 I think its time that the quicklook thing gets sorted out. If its possible please release a new version of quicklook just for the flow launcher to work with. The always preview option is pretty janky with the plugin turned on. At the very least keep the inbuilt preview "on" at all times and let the quicklook preview open with a cusomizable keybinding.

**Comment by VictoriousRaptor** [comment]
@jjw24 The always preview part is what we've discussed before this plugin was rolled out. Should we consider ignoring always preview in QL plugin?

**Comment by jjw24** [comment]
>@jjw24 The always preview part is what we've discussed before this plugin was rolled out. Should we consider ignoring always preview in QL plugin?

I am ok with this, @nanoraptor ?

>QL doesn't provide an interface so Flow cannot properly detect it's status. What about detecting QL window presence when changing preview state?

Thats a good idea. Add a method for the plugin to provide status feedback (if one doesnt already exist), detect status by detecting QL window presence.

**Comment by nanoraptor** [comment]
@jjw24 I'm ok with it as well. Always preview can be ignored in QL Plugin. Nobofy wants to see the preview if they can't even see what they've selected. Make your default preview to be the "default". If someone wants to use ql they can do so by pressing a hotkey. It should be an additional layer over your default one. It should bot replace the default preview!
